sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+             (uint64_t)b_in(emmc, BCMEMMC_RESP1) << 32;
+      return 0;
+    case EMMC_PROP_RW_RESP_HI:
+      *var = (uint64_t)b_in(emmc, BCMEMMC_RESP2) |
+             (uint64_t)b_in(emmc, BCMEMMC_RESP3) << 32;
+      return 0;
+    case EMMC_PROP_RW_BUSWIDTH:
+      *var = bcmemmc_get_bus_width(state);
+      return 0;
+    case EMMC_PROP_RW_RCA:
+      *var = state->rca;
+      return 0;
+    default:
+      return ENODEV;
+  }
+}
+
+static int bcmemmc_set_prop(device_t *cdev, uint32_t id, uint64_t var) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+  resource_t *emmc = state->emmc;
+
+  uint32_t reg = 0;
+  switch (id) {
+    case EMMC_PROP_RW_BLKCNT:
+      if (var & ~0xffff)
+        return EINVAL;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT);
+      reg = (reg & 0x0000ffff) | (var << 16);
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg);
+      return 0;
+    case EMMC_PROP_RW_BLKSIZE:
+      if (var & ~0x03ff)
+        return 0;
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT);
+      reg = (reg & ~0x03ff) | var;
+      b_out(emmc, BCMEMMC_BLKSIZECNT, reg);
+      return 0;
+    case EMMC_PROP_RW_RESP_LOW:
+      b_out(emmc, BCMEMMC_RESP0, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP1, (uint32_t)(var >> 32));
+      return 0;
+    case EMMC_PROP_RW_RESP_HI:
+      b_out(emmc, BCMEMMC_RESP2, (uint32_t)var);
+      b_out(emmc, BCMEMMC_RESP3, (uint32_t)(var >> 32));
+      return 0;
+    case EMMC_PROP_RW_CLOCK_FREQ:
+      return bcmemmc_clk(cdev->parent, var);
+    case EMMC_PROP_RW_BUSWIDTH:
+      return bcmemmc_set_bus_width(state, var);
+    case EMMC_PROP_RW_RCA:
+      state->rca = var;
+      return 0;
+    default:
+      return ENODEV;
+  }
+}
+
+static int bcmemmc_cmd_code(device_t *dev, uint32_t code, uint32_t arg,
+                            emmc_resp_t *resp) {
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+
+  uint32_t r = 0;
+
+  b_out(emmc, BCMEMMC_ARG1, arg);
+  b_out(emmc, BCMEMMC_CMDTM, code);
+  if ((r = bcmemmc_intr_wait(dev, INT_CMD_DONE))) {
+    klog("ERROR: failed to send EMMC command %p", code);
+    return r;
+  }
+
+  if (resp) {
+    resp->r[0] = b_in(emmc, BCMEMMC_RESP0);
+    resp->r[1] = b_in(emmc, BCMEMMC_RESP1);
+    resp->r[2] = b_in(emmc, BCMEMMC_RESP2);
+    resp->r[3] = b_in(emmc, BCMEMMC_RESP3);
+  }
+
+  return 0;
+}
+
+static int bcmemmc_cmd(device_t *cdev, emmc_cmd_t cmd, uint32_t arg,
+                       emmc_resp_t *resp) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+
+  if (cmd.flags & EMMC_F_APP)
+    bcmemmc_cmd(cdev, EMMC_CMD(APP_CMD), state->rca << 16, NULL);
+
+  uint32_t code = encode_cmd(cmd);
+  return bcmemmc_cmd_code(cdev->parent, code, arg, resp);
+}
+
+static int bcmemmc_read(device_t *cdev, void *buf, size_t len, size_t *read) {
```

Comment:
It seams that whenever we want to return a value through an argument, the convention is to finish the pointer's name with `_p` (@cahirwpz correct me if I'm wrong).

Reply:
I can do that, but I'm not really sure if that's the case.

Reply:
On the other hand, why won't you just use `ssize_t` like in the `read` `write` syscalls.

Reply:
Why would I? This is not a syscall. Does it make sense to have a signed type here?